### DISCUSSION TITLE
ALTER DEFAULT PRIVILEGES REVOKE EXECUTE ON FUNCTIONS FROM PUBLIC; wont revoke anon & authenticated function execution

### DISCUSSION THREAD

**Discussion by ShanArosh** [supabase/supabase#15809] [Questions] [answered]
I used supabase official [Function privileges](https://supabase.com/docs/guides/database/functions#function-privileges) documentation. and then used the client library to send requests to rest API as anon & authenticated to test if revoke worked as expected. but clearly, it isn't. Is this a bug or I'm doing something wrong?

```
ALTER DEFAULT PRIVILEGES REVOKE EXECUTE ON FUNCTIONS FROM PUBLIC;
```

```
create or replace function public.try()
returns text
language plpgsql
as $$
begin
  return 'try 1 ' || current_user ||' ' || current_role;
end;
$$;
```
```
ALTER DEFAULT PRIVILEGES REVOKE EXECUTE ON FUNCTIONS FROM PUBLIC;

-- Choose which roles can execute functions
GRANT EXECUTE ON FUNCTION public.try() TO service_role;

```

**Comment by GaryAustin1** [comment]
I believe the documentation will only impact new functions.
For an existing function you need to do:
`revoke all privileges on function my_function from anon, authenticated;`

**Reply by ShanArosh** [reply]
I ran the revoke `ALTER DEFAULT PRIVILEGES REVOKE EXECUTE ON FUNCTIONS FROM PUBLIC;` before the creation of the function. Any idea what this might be ?

**Reply by GaryAustin1** [reply, marked as answer]
The functions by default were granted anon,authenticated,service_role,postgres directly by Supabase setup if I recall.
So you need to revoke anon and authenticated instead of public.  The docs might be wrong on using public. 
But I've not tried this.

This is the example they give for restoring the default privileges if they get messed up.
![image](https://github.com/supabase/supabase/assets/54564956/297677dd-41b2-4739-bc51-63a3a4eb31e4)

**Reply by ShanArosh** [reply]
Thank you @GaryAustin1 for the quick reply.
